include/swift/Driver/ExperimentalDependencyDriverGraph.h

```diff
@@ -0,0 +1,338 @@
+using DependencyKey = swift::experimental_dependencies::DependencyKey;
+using DeclAspect = swift::experimental_dependencies::DeclAspect;
+
+template <typename Key1, typename Key2, typename Value>
+using TwoStageMap =
+    swift::experimental_dependencies::TwoStageMap<Key1, Key2, Value>;
+
+template <typename Key1, typename Key2, typename Value>
+using BiIndexedTwoStageMap =
+    swift::experimental_dependencies::BiIndexedTwoStageMap<Key1, Key2, Value>;
+
+using DeclAspect = swift::experimental_dependencies::DeclAspect;
+
+//==============================================================================
+// MARK: DriverNode
+//==============================================================================
+
+/// A node in the DriverDependencyGraph
+/// Keep separate type from Node for type-checking.
+class DriverNode : public Node {
+
+public:
+  DriverNode(const DependencyKey &key, Optional<std::string> fingerprint)
+      : Node(key, fingerprint) {}
+
+  /// Integrate \p integrand's fingerprint into \p dn.
+  /// \returns true if there was a change requiring recompilation.
+  bool integrateFingerprintFrom(const FrontendNode *integrand) {
+    if (getFingerprint() == integrand->getFingerprint())
+      return false;
+    setFingerprint(integrand->getFingerprint());
+    return true;
+  }
+};
+
+/// A placeholder allowing the experimental system to fit into the driver
+/// without changing as much code.
+class DependencyGraphImpl {
+public:
+  /// Use the status quo LoadResult for now.
+  using LoadResult = typename swift::DependencyGraphImpl::LoadResult;
+};
+
+//==============================================================================
+// MARK: DriverGraph
+//==============================================================================
+
+/// See \ref Node in ExperimentalDependencies.h
+class DriverGraph {
+
+  /// Find nodes, first by the swiftDeps file, then by key.
+  /// Supports searching specific files for a node matching a key.
+  /// Such a search is useful when integrating nodes from a given source file to
+  /// see which nodes were there before integration and so might have
+  /// disappeared.
+  ///
+  /// Some nodes are in no file, for instance a dependency on a Decl in a source
+  /// file whose swiftdeps has not been read yet. For these, the filename is the
+  /// empty string.
+  ///
+  /// Don't add to this collection directly; use \ref addToMap
+  /// instead because it enforces the correspondence with the swiftFileDeps
+  /// field of the node.
+  /// TODO: Fix above comment
+  ///
+  /// Sadly, cannot use an optional string for a key.
+  using NodeMap =
+      BiIndexedTwoStageMap<std::string, DependencyKey, DriverNode *>;
+  NodeMap nodeMap;
+
+  /// Since dependency keys use baseNames, they are coarser than individual
+  /// decls. So two decls might map to the same key. Index the second level by
+  /// file. Very useful to find a Node given the key, for instance when finding
```

Comment:
I'm not sure what "index the second level by file" means.

Reply:
Great feedback. Took a stab at a reword.